enhancements/network/nodeport-range-configuration.md

```diff
@@ -0,0 +1,125 @@
+---
+title: Make Service NodePort Range Configurable
+authors:
+  - "@abhat"
+reviewers:
+  - "@dcbw"
+  - "@deads2k"
+  - "@smarterclayton"
+  - "@danwinship"
+  - "@knobunc"
+approvers:
+  - "@knobunc"
+creation-date: 2020-07-06
+last-updated: 2020-07-06
+status: implementable
+---
+# Configurability of Service NodePort Range
+This enhancement targets making the service node-port range of an Openshift cluster configurable.
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Service node-port range is a [configurable parameter](https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/bindata/v4.1.0/config/defaultconfig.yaml#L147)
+on the Kubernetes API server. As our customers scale the number of node-port services in their Openshift clusters,
+there is a desire to make this node port range configurable.
+
+## Motivation
+
+Service node-port range is a configurable parameter consumed by the Kubernetes API server as
+a slice of port ranges. It defaults currently to a single port range (30000-32767).
+It is consumed in Openshift via the ConfigMap resource that holds all the api-server configuration.
+In order to allow customers to scale the number of node-port services in the cluster to a value greater
+than that allowed by the default service node-port range, it is important to allow dynamically changing this
+range to allow for more ports. This enhancement focuses on the mechanics of allowing users to configure a
+custom service node-port range for their clusters.
+
+### Goals
+
+- Customers can configure the service node-port range for their cluster during install.
+- Customers can configure the service node-port range for their cluster post install as their needs change.
+
+### Non-Goals
+
+- Services launched prior to customers changing the service node-port range will not work if the new range is
+  not inclusive of the old range.
+- This enhancement makes no guarantees of providing non-disruptive change to the service node-port range.
+  In other words customers, when they make non-inclusive changes to the service node-port range are responsible
+  for deleting old services and recreating new services that will then get node ports assigned from the
+  new range.
+- Configuration of multiple service node-port ranges even though supported by the Kubernetes API server will not be
```

Comment:
Why?

Reply:
This was an assumption made for simplicity. If there is value in supporting disjoint ranges, we can add this easily.

Reply:
If a customer ever asks us to support it, and we decide to, starting with single range may complicate the problem because now we will have to change the API.  If multiple range is what Kube supports, and supporting mulitple doesn't significantly increase the total scope of work (dev test doc) I would prefer to do multiple range.

Reply:
the portallocator uses a bitmap, snapshoted in etcd, with the size of the port range. 
If we want to support multiple ranges we have to use multiple bitmaps, no?

Reply:
Should we just make the api take a list of ranges, but document (and throw an error) that only one range is supported.

Reply:
> the portallocator uses a bitmap, snapshoted in etcd, with the size of the port range.
> If we want to support multiple ranges we have to use multiple bitmaps, no?

That is an implementation detail of the API Server no? Our configuration surface is the `network.config.openshift.io` CRDs `cluster` resource.

Reply:
Also, a quick clarification. Looks like the Kubernetes API server only allows a single range. But the default config we have here: https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/bindata/v4.1.0/config/defaultconfig.yaml#L147 allows for a slice which is simply a map[string][]string. So while the default config looks like it can support multiple ranges, in practice kubernetes only supports a single range. With that background, do we need to support multiple ranges in `openshift/api`?

Reply:
@aojea mentioned in standup that for the IPv6 work the community deliberately avoided a second allocator, so it looks unlikely that they will support multiple ranges any time soon.  So it looks like we can safely just use a single range.

Reply:
just for reference https://github.com/kubernetes/kubernetes/issues/90101#issuecomment-660691340